flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherCleanupITCase.java

```diff
@@ -179,6 +188,60 @@ public void testCleanupThroughRetries() throws Exception {
                 "The JobResultStore should have this job marked as clean.");
     }
 
+    @Test
+    public void testCleanupNotCancellable() throws Exception {
+        final JobGraph jobGraph = createJobGraph();
+        final JobID jobId = jobGraph.getJobID();
+
+        final JobResultStore jobResultStore = new EmbeddedJobResultStore();
+        jobResultStore.createDirtyResult(
+                new JobResultEntry(TestingJobResultStore.createSuccessfulJobResult(jobId)));
+        haServices.setJobResultStore(jobResultStore);
+
+        // Instantiates JobManagerRunner
+        final CompletableFuture<Void> jobManagerRunnerCleanupFuture = new CompletableFuture<>();
+        final AtomicReference<JobManagerRunner> jobManagerRunnerEntry = new AtomicReference<>();
+        final JobManagerRunnerRegistry jobManagerRunnerRegistry =
+                TestingJobManagerRunnerRegistry.builderFromSingleEntry(jobManagerRunnerEntry)
+                        .withGlobalCleanupAsyncFunction(
+                                (actualJobId, executor) -> jobManagerRunnerCleanupFuture)
+                        .build();
+
+        final Dispatcher dispatcher =
+                createTestingDispatcherBuilder()
+                        .setJobManagerRunnerRegistry(jobManagerRunnerRegistry)
+                        .build();
+        dispatcher.start();
+
+        toTerminate.add(dispatcher);
+
+        CommonTestUtils.waitUntilCondition(
+                () -> jobManagerRunnerEntry.get() != null,
+                Deadline.fromNow(Duration.ofSeconds(10)),
+                "JobManagerRunner wasn't loaded in time.");
+
+        assertThat(
+                "The JobResultStore should have this job still marked as dirty.",
+                haServices.getJobResultStore().hasDirtyJobResultEntry(jobId),
+                CoreMatchers.is(true));
+
+        final DispatcherGateway dispatcherGateway =
+                dispatcher.getSelfGateway(DispatcherGateway.class);
+
+        try {
+            dispatcherGateway.cancelJob(jobId, TIMEOUT).get();
+            Assert.fail("Should fail because cancelling the cleanup is not allowed.");
+        } catch (Throwable e) {
+            assertThat(e, FlinkMatchers.containsCause(FlinkException.class));
```

Comment:
nit: it would be great if we could also assert the exception message here, as the `FlinkException` could hide lot of things (I've recently had a hard time debugging an instability because of this -> the exception meant something else)

Reply:
good point. I introduced a new `JobCancellationFailedException` to cover this specific use-case.